pkg/nexus/nexus.go

```diff
@@ -53,53 +54,125 @@ func (n *BinaryDownloader) DownloadArtifact(c *config.MavenGav) (Deliverable, er
-	if err != nil {
-		return deliverable, errors.Wrap(err, "Failed to parse content-disposition")
-	}
-	dir, err := ioutil.TempDir("", "package")
-	if err != nil {
-		return deliverable, errors.Wrap(err, "Failed to create directory for artifact")
-	}
-	fileName := filepath.Join(dir, params["filename"])
+	deliverable := Deliverable{}
 
-	fileCreated, err := os.Create(fileName)
+	// Detect API version
+	useNexus3 := false
+	resp, err := http.Get(n.baseUrl)
 	if err != nil {
-		return deliverable, errors.Wrap(err, "Failed to create artifact file")
+		logrus.Fatalf("Failed response from %s, error: %s", n.baseUrl, err)
+	} else {
+		rx := regexp.MustCompile(`Nexus/\d`)
+		nexusVer := rx.FindString(resp.Header.Get("Server"))
+		logrus.Infof("Nexus is %s", nexusVer)
+		useNexus3 = string(nexusVer) == "Nexus/3"
 	}
-	defer fileCreated.Close()
 
-	_, err = io.Copy(fileCreated, httpResponse.Body)
-	if err != nil {
-		return deliverable, errors.Wrap(err, "Failed to write to artifact file")
+	if useNexus3 {
+		// Nexus 3
+		resourceUrl, err := n.createNexus3URL(c)
+		if err != nil {
+			return deliverable, errors.Wrapf(err, "Failed to create Nexus 3 url for GAV %+v", c)
+		}
+		logrus.Debugf("Downloading artifact from %s", resourceUrl)
+
+		if na == nil || na.Username == "" || na.Password == "" {
+			return deliverable, errors.Wrap(err, "Missing Nexus credentials for Nexus 3")
+		}
+
+		req, err := http.NewRequest("GET", resourceUrl, nil)
+		if err != nil {
+			return deliverable, errors.Wrapf(err, "Failed to create request for Nexus url %s", resourceUrl)
+		}
+		req.SetBasicAuth(na.Username, na.Password)
+
+		httpResponse, err := http.DefaultClient.Do(req)
+		if err != nil {
+			return deliverable, errors.Wrapf(err, "Failed to get artifact from Nexus %s", resourceUrl)
+		}
+		defer httpResponse.Body.Close()
+
+		if httpResponse.StatusCode != http.StatusOK {
+			return deliverable, errors.Errorf("Could not download artifact (Make sure you have deployed it!)"+
+				". Status code %s ", httpResponse.Status)
+		}
+
+		dir, err := ioutil.TempDir("", "package")
+		if err != nil {
+			return deliverable, errors.Wrap(err, "Failed to create directory for artifact")
+		}
+		// No content-disposition from nexus3 - using alternate filename composition
+		if string(c.Classifier) == "" {
+			return deliverable, errors.Errorf("Missing maven Classifier")
+		}
+		if string(c.Type) == "" {
+			return deliverable, errors.Errorf("Missing maven Type")
+		}
+		fileName := filepath.Join(dir, string(c.Classifier)+"."+string(c.Type))
+
+		fileCreated, err := os.Create(fileName)
+		if err != nil {
+			return deliverable, errors.Wrap(err, "Failed to create artifact file")
+		}
+		defer fileCreated.Close()
+
+		_, err = io.Copy(fileCreated, httpResponse.Body)
+		if err != nil {
+			return deliverable, errors.Wrap(err, "Failed to write to artifact file")
+		}
+		deliverable.Path = fileName
+		logrus.Debugf("Downloaded artifact to %s", deliverable.Path)
+	} else {
```

Comment:
Her er det duplisering av kode.

Det som varierer er etter hva jeg kan se:

* URL
* Filnavn som det skal lagres til

Brukernavn/passord kan gjøres likt. Er det ikke satt, sendes det ikke.

Er det mulig å slå sammen de to kodeblokkene?

Reply:
Jeg prøvde i starten uten hell, men forsøker igjen. Koden har endret seg litt siden første forsøk.

Reply:
Jeg tenker det samme. Og koden for filnavn bør kunne brukes for ny/gammel Nexus. Ellers syns alle feil ved nedlasting bør inneholde URL til ressursen, gjør det litt mer transparent.

Reply:
Har forsøkt å slå sammen nå.  Vet ikke om jeg synes det ble mer lesbart.  Med tanke på å rydde vekk Nexus 2-spesifikk kode senere, så foretrakk jeg det med duplisering i to distinkte kodeblokker.

Reply:
Foreslår å trekke de to jeg kommenterte "metode" på ut i egne metoder. Tror det kan hjelpe på lesbarheten.

Ellers så syns jeg det er veldig mye lettere å forholde meg til en flyt, så syns endringen er til det bedre.

Reply:
Trekker ut metoder.